incubator/hnc/pkg/reconcilers/hierarchy_config.go

```diff
@@ -184,31 +174,60 @@ func (r *HierarchyConfigReconciler) onMissingNamespace(log logr.Logger, ns *fore
 		return false
 	}
 
+	// If the namespace is removed, no more action need on the reconciliation. The
+	// namespace is only unset existence below.
 	if !ns.Exists() {
-		// The namespace doesn't exist on the server, but its owner expects it to be there. Initialize
-		// it so it gets created; once it is, it will be reconciled again.
-		if ns.Owner != "" {
-			log.Info("Will create missing namespace", "forOwner", ns.Owner)
-			nsInst.Name = ns.Name()
-			// Set "api.AnnotationOwner" annotation to the non-existent yet namespace.
-			metadata.SetAnnotation(nsInst, api.AnnotationOwner, ns.Owner)
-		}
 		return true
 	}
 
 	// Remove it from the forest and notify its relatives
 	r.enqueueAffected(log, "relative of deleted namespace", ns.RelativesNames()...)
-	// Enqueue the HNS if the owned namespace is deleted.
-	if r.HNSReconcilerEnabled {
-		if ns.Owner != "" {
-			r.hnsr.enqueue(log, ns.Name(), ns.Owner, "hns for the deleted owned namespace")
-		}
-	}
 	ns.UnsetExists()
 	log.Info("Removed namespace")
 	return true
 }
 
+// syncOwner sets the parent to the owner and updates the HNS_MISSING condition
+// if the HNS instance is missing in the owner namespace according to the forest.
+// The namespace owner annotation is the source of truth of the ownership, since
+// modifying a namespace has higher privilege than what HNC users can do.
+func (r *HierarchyConfigReconciler) syncOwner(log logr.Logger, inst *api.HierarchyConfiguration, nsInst *corev1.Namespace, ns *forest.Namespace) {
+	// Clear the HNS_MISSING condition if this is not an owned namespace or to
+	// reset it for the updated condition later.
+	ns.ClearAllConditions(api.HNSMissing)
+	nm := ns.Name()
+	onm := nsInst.Annotations[api.AnnotationOwner]
+	ons := r.Forest.Get(onm)
+
+	if onm == "" {
+		return
+	}
+
+	if ns.Owner != onm {
+		log.Info("Owner annotation was updated", "old", ns.Owner, "new", onm)
+		ns.Owner = nsInst.Annotations[api.AnnotationOwner]
+	}
+
+	if inst.Spec.Parent != onm {
+		if ns.Parent() == nil && inst.Spec.Parent == "" {
+			// This is the first time an owned namespace is reconciled in HCR. We need
+			// to enqueue the HNS so it can update the state from 'Missing' to 'Ok'.
+			r.hnsr.enqueue(log, nm, onm, "the missing owned namespace is found")
+		}
```

Comment:
Ah right, I should put the `r.hnsr.enqueue()` to the `markExisting()` when `ns.Owner != ""`. We still need this `hnsr.enqueue` since the `nsInst` doesn't change in this hcr, thus it won't trigger the hnsr by watch naturally, right?

Reply:
Right, sorry, I misread `markExisting`. Your new code lgtm.